### ISSUE TITLE
Add warning re unsaved changes in Component modal

### ISSUE THREAD

**Issue by johnmhoran** [nexB/aboutcode-manager#127]
In the Component modal, we have the choice of “Close”, “Delete Component” and “Save changes”.  If you make an edit and click “Close”, the popup closes and any edits are lost -- there's no warning that you’ll lose your unsaved changes.  A warning might save a user from inadvertently losing his or her carefully-considered edits.

**Comment by johnmhoran** [comment]
Hi @jdaguil . Just wanted to touch base re my status -- I'm currently working on this issue. 

I began by testing whether the relevant form fields were empty or not, and then realized that editing an existing Component would have some fields already populated, so testing for empty fields is not sufficient.  

My research led me to the jQuery `.serialize()` method, and I'm in the process of trying to figure out how to deploy it inside `componentDialog.js`, find the form we're interested in (there are several in `index.html` so I've given the target an `id`), serialize the contents and then compare that with the value retrieved when the close event begins.

I've not yet been able to find the target form and run the `serialize()` method on it -- the structure of our `ComponentDialog` class differs from the various jQuery `serialize()` docs, examples etc. out there in the wild.  I'm also not yet sure how to retrieve the pre-close state of the form against which the serialization will be compared.  

I'll keep you apprised of my progress, and as always would welcome any suggestions you might have.

**Comment by johnmhoran** [comment]
Curiously, this serializes one of the forms:  `let str = $('form' ).serialize();` and in the console displays the expected data: `product=&version=&version=&api=` -- but for the wrong form, in this case the last form in `index.html`.

Using the `id` I've added to our target form, this: `let str = $('#nodeForm').serialize();` appears to be empty.

**Comment by johnmhoran** [comment]
One other clue: `let form = this.dialog.find("#nodeForm");` finds the target form and when displayed in the console comprises an outline structure that can be opened node-by-node and even includes the default values of the various form contents.

**Comment by johnmhoran** [comment]
@jdaguil It may be relevant that the form I *can* serialize consists of a series of `divs` containing label/input and label/select pairs.  The target form, in contrast, places those pairs inside a `fieldset` rather than a `div`.

**Comment by jdaguil** [comment]
It sounds like you were able to get the correct form. One suggestion would be to just use 
```
this.dialog.find("form")
```
You don't need to specify the exact `id` bc you are searching within the dialog and we know there's only one form.

With regard to serialization, I think the problem is that your missing the `name` attribute on the inputs. From the [serialize documentation](https://api.jquery.com/serialize/):

>  For a form element's value to be included in the serialized string, the element must have a name attribute.

**Comment by johnmhoran** [comment]
@jdaguil  After testing, I've concluded that using serialization is not the correct approach, given the cumbersome construction of the generated string.  This is an example of the serialization output (representing the state of the form's `inputs`, `selects` etc. just before the user attempts to close the dialog):
```
component-status=Attention&component-name=Test+Component+001&component-version=1.0.2&component-license=boost-1.0&component-license=mit-open-group&component-owner=Chris+Anderson&component-copyright=(c)+1995-2017+Jean-loup+Gailly+%26+Mark+Adler&component-language=Ada&component-notes=These+are+my+notes.
```
Note among other things how the user's choice of two licenses (`boost-1.0` and `mit-open-group`) is represented.

Here's the approach I've taken (will commit and push shortly):

* In `index.html`, inside the Node View Modal, add an `id` to the two buttons available to close the modal (`Close` at the bottom and the `X` in the upper right hand corner).
* In `componentDialog.js`:
  * Define these two buttons as attributes (or is the correct term "properties"?) inside the `ComponentDialog` class constructor and connect their respective `click` events to a new method named `_closeComponent()`.
  * In the constructor, define attributes -- one for each of the form's `inputs` etc. -- that will be modified as necessary in `_showDialog()` to represent any previously-saved values.
  * In `_showDialog()`, disable the ability to close the dialog by clicking outside the dialog or pressing the escape key, and set the value of each of our new attributes equal to the previously-saved value (if any).
  * In the new method `_closeComponent()`, test whether any of the previously-saved values has been changed without saving, display an alert if a change is detected, and in the alert, return `false` (leaving the modal/dialog open) if the user clicks `Cancel`.  Note that the four values represented as arrays rather than strings (`component-license`, `component-owner`, `component-copyright` and `component-language`) are tested using the `.not` selector rather than with simple equality.

Looking forward to your comments and suggestions.

**Comment by johnmhoran** [comment]
@jdaguil All our tests pass and everything looks good, with one exception -- after building the `dist` target and creating and saving a new Component for a `zlib` scan, I navigated to the `Component Summary` view and clicked `Upload Components`, which threw the following error in the console:
```
Unhandled rejection TypeError: dejaCodeExportDialog is not a function
```
This does not appear to be related to my work on this issue but I'm not certain of that.  Do you have any thoughts?  Seems to be triggered here: `assets/app/js/componentDataTables.js:59:60`, inside the `_createDataTable()` method.

I'm going to go ahead and push my commit so you can vet, but won't open a PR until I get your go-ahead.

**Comment by johnmhoran** [comment]
@jdaguil I just checked out the local `develop` branch, did a build and ran the same test on a `zlib` scan, getting the same error message.  Will update `develop` (my local `develop` is 2 commits behind `origin/develop`), test and report.

**Comment by johnmhoran** [comment]
After updating `develop`, running our protocol and building a new `dist` -- same error message: 
```
Unhandled rejection TypeError: dejaCodeExportDialog is not a function 
```

**Comment by jdaguil** [comment]
@johnmhoran thanks for the report. This is something that is in my plan to fix with my work in #168. It broke with some of the refactoring. Please feel free to open a PR for your work when you're ready.
